### PR TITLE
initial support for npm

### LINKED ISSUE #16: Add package.info and README.md files for NPM packages
As each distribution should be published on NPM, we need the `package.json` and `readme.md` files in each distribution to be able to publish them on NPM after the next release. 

A note about this should also be added to the release process document

### PR DESCRIPTION
**Pull request by pmlopes**
Could you maven experts give me some review on this?

The idea is to have an automated way to deploy to NPM

@cescoffier @vietj 

Cheers,
Paulo

(Fix #16)

### CODE REVIEW

File: stack-npm/pom.xml [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>io.vertx</groupId>
+    <artifactId>vertx-stack</artifactId>
+    <version>3.1.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>vertx-stack-npm</artifactId>
+  <version>3.1.0-SNAPSHOT</version>
```
Comment (cescoffier) [inline stack-npm/pom.xml]: Remove it (will be inherited from parent)

File: stack-npm/pom.xml [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+  <properties>
+    <asciidoc.dir>${project.basedir}/src/main/asciidoc</asciidoc.dir>
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>io.vertx</groupId>
+      <artifactId>vertx-stack-dist</artifactId>
+      <version>3.1.0-SNAPSHOT</version>
```
Comment (cescoffier) [inline stack-npm/pom.xml]: It's a question of taste, but we generally use ${project.version}

File: stack-npm/pom.xml [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+                      <filtering>true</filtering>
+                    </resource>
+                  </resources>
+                </configuration>
+              </execution>
+            </executions>
+          </plugin>
+          <plugin>
+            <artifactId>exec-maven-plugin</artifactId>
+            <groupId>org.codehaus.mojo</groupId>
```
Comment (cescoffier) [inline stack-npm/pom.xml]: We should use the maven frontend plugin to run npm. 
See https://github.com/vert-x3/vertx-web/commit/a999841ff1b66cee59456576e3bbc957c1ebc6b1#diff-600376dffeb79835ede4a0b285078036R200

File: stack-npm/pom.xml [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                </configuration>
+              </execution>
+            </executions>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
+
+    <profile>
+      <!-- Use -DskipDoc to avoid building doc -->
```
Comment (cescoffier) [inline stack-npm/pom.xml]: Probably useless as we don't have 'documentation' (in the vert.x meaning)

File: stack-npm/pom.xml [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+  </parent>
+
+  <artifactId>vertx-stack-npm</artifactId>
+  <version>3.1.0-SNAPSHOT</version>
+
+  <name>Vert.x Stack - NPM packages</name>
+  <packaging>pom</packaging>
+
+  <properties>
+    <asciidoc.dir>${project.basedir}/src/main/asciidoc</asciidoc.dir>
```
Comment (cescoffier) [inline stack-npm/pom.xml]: Is it useful ?

**Comment by cescoffier** [comment]
Except if you plan to write a manual, it's probably better to remove all references to asciidoc. 
It should use the maven frontend plugin to execute NPM as it:

1. provisions Node and NPM
2. behavior the same way on all OS (windows included)

**Comment by cescoffier** [comment]
For information - working on the branch.

**Comment by cescoffier** [comment]
Made some changes:

* Add the vertxbus.js NPM here (initially made here: https://github.com/vert-x3/vertx-web/issues/137)
* Refactoring of the build
* Extraction of the deployment to its own profile

@vietj @pmlopes : ready for review



The release process has also been updated:
https://github.com/vert-x3/wiki/wiki/Release-Process

**Comment by vietj** [comment]
looks good to me

**Comment by pmlopes** [comment]
:+1:

**Pull request merged**